File: app/src/main/java/com/sjl/bookmark/ui/presenter/ArticleSearchPresenter.java
package com.sjl.bookmark.ui.presenter;

import com.sjl.bookmark.api.WanAndroidApiService;
import com.sjl.bookmark.entity.Article;
import com.sjl.bookmark.entity.DataResponse;
import com.sjl.bookmark.entity.HotKey;
import com.sjl.bookmark.ui.contract.ArticleSearchContract;
import com.sjl.core.net.RetrofitHelper;
import com.sjl.core.net.RxSchedulers;
import com.sjl.core.util.log.LogUtils;

import java.util.List;

import io.reactivex.functions.Consumer;

/**
 * TODO
 *
 * @author Kelly
 * @version 1.0.0
 * @filename ArticleSearchPresenter.java
 * @time 2018/4/1 12:03
 * @copyright(C) 2018 song
 */
public class ArticleSearchPresenter extends ArticleSearchContract.Presenter {

    private int mCurrentPage;



    /**
     * 获取热门搜索关键字
     */
    @Override
    public void getHotKeyData() {
        WanAndroidApiService apiService = RetrofitHelper.getInstance().getApiService(WanAndroidApiService.class);
        apiService.getHotKeys()
                .compose(RxSchedulers.<DataResponse<List<HotKey>>>applySchedulers()).as(this.<DataResponse<List<HotKey>>>bindLifecycle())
                .subscribe(new Consumer<DataResponse<List<HotKey>>>() {
                    @Override
                    public void accept(DataResponse<List<HotKey>> dataResponse) throws Exception {
                        mView.getHotKeySuccess(dataResponse.getData());
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        mView.showFailMsg("热门搜索：" + throwable.getMessage());
                    }
                });
    }

    /**
     * 根据关键字搜索
     * @param key
     */
    @Override
    public void searchData(String key) {
        mCurrentPage = 0;
        WanAndroidApiService apiService = RetrofitHelper.getInstance().getApiService(WanAndroidApiService.class);

        apiService.getSearchArticles(mCurrentPage, key)
                .compose(RxSchedulers.<DataResponse<Article>>applySchedulers()).as(this.<DataResponse<Article>>bindLifecycle())
                .subscribe(new Consumer<DataResponse<Article>>() {
                    @Override
                    public void accept(DataResponse<Article> dataResponse) throws Exception {
                        mView.searchDataSuccess(dataResponse.getData().getDatas());
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        LogUtils.e(throwable);
                        mView.showFailMsg("搜索：" + throwable.getMessage());
                    }
                });
    }

    /**
     * 上拉加载更多数据
     * @param keyWord
     */
    @Override
    public void getMoreData(String keyWord) {
        mCurrentPage += 1;
        WanAndroidApiService apiService = RetrofitHelper.getInstance().getApiService(WanAndroidApiService.class);

        apiService.getSearchArticles(mCurrentPage, keyWord)
                .compose(RxSchedulers.<DataResponse<Article>>applySchedulers()).as(this.<DataResponse<Article>>bindLifecycle())
                .subscribe(new Consumer<DataResponse<Article>>() {
                    @Override
                    public void accept(DataResponse<Article> dataResponse) throws Exception {
                        mView.loadMoreDataSuccess(dataResponse.getData().getDatas());
                    }
                }, new Consumer<Throwable>() {
                    @Override
                    public void accept(Throwable throwable) throws Exception {
                        mView.showFailMsg("分页加载：" + throwable.getMessage());
                    }
                });

    }
}

File: app/src/main/java/com/sjl/bookmark/ui/presenter/ArticleSearchPresenter.kt
package com.sjl.bookmark.ui.presenter

import com.sjl.bookmark.R
import com.sjl.bookmark.api.WanAndroidApiService
import com.sjl.bookmark.entity.Article
import com.sjl.bookmark.entity.DataResponse
import com.sjl.bookmark.entity.HotKey
import com.sjl.bookmark.ui.contract.ArticleSearchContract
import com.sjl.core.net.RetrofitHelper
import com.sjl.core.net.RxSchedulers
import com.sjl.core.util.log.LogUtils
import com.uber.autodispose.ObservableSubscribeProxy
import io.reactivex.functions.Consumer

/**
 * TODO
 *
 * @author Kelly
 * @version 1.0.0
 * @filename ArticleSearchPresenter.java
 * @time 2018/4/1 12:03
 * @copyright(C) 2018 song
 */
class ArticleSearchPresenter : ArticleSearchContract.Presenter() {
    private var mCurrentPage: Int = 0

    /**
     * 获取热门搜索关键字
     */
    override fun getHotKeyData() {
        val apiService: WanAndroidApiService = RetrofitHelper.getInstance().getApiService(
            WanAndroidApiService::class.java
        )
        apiService.hotKeys
            .compose(RxSchedulers.applySchedulers())
            .`as`<ObservableSubscribeProxy<DataResponse<List<HotKey>>>>(
                bindLifecycle<DataResponse<List<HotKey>>>()
            )
            .subscribe(object : Consumer<DataResponse<List<HotKey>>> {
                @Throws(Exception::class)
                override fun accept(dataResponse: DataResponse<List<HotKey>>) {
                    mView.getHotKeySuccess(dataResponse.data)
                }
            }, object : Consumer<Throwable> {
                @Throws(Exception::class)
                override fun accept(throwable: Throwable) {
                    mView.showFailMsg("热门搜索：" + throwable.message)
                }
            })
    }

    /**
     * 根据关键字搜索
     * @param key
     */
    override fun searchData(key: String) {
        mCurrentPage = 0
        val apiService: WanAndroidApiService = RetrofitHelper.getInstance().getApiService(
            WanAndroidApiService::class.java
        )
        apiService.getSearchArticles(mCurrentPage, key)
            .compose(RxSchedulers.applySchedulers())
            .`as`<ObservableSubscribeProxy<DataResponse<Article>>>(
                bindLifecycle<DataResponse<Article>>()
            )
            .subscribe(object : Consumer<DataResponse<Article>> {
                @Throws(Exception::class)
                override fun accept(dataResponse: DataResponse<Article>) {
                    mView.searchDataSuccess(dataResponse.data.datas)
                }
            }, object : Consumer<Throwable> {
                @Throws(Exception::class)
                override fun accept(throwable: Throwable) {
                    LogUtils.e(throwable)
                    mView.showFailMsg("搜索：" + throwable.message)
                }
            })
    }

    /**
     * 上拉加载更多数据
     * @param keyWord
     */
    override fun getMoreData(keyWord: String) {
        mCurrentPage += 1
        val apiService: WanAndroidApiService = RetrofitHelper.getInstance().getApiService(
            WanAndroidApiService::class.java
        )
        apiService.getSearchArticles(mCurrentPage, keyWord)
            .compose(RxSchedulers.applySchedulers())
            .`as`<ObservableSubscribeProxy<DataResponse<Article>>>(
                bindLifecycle<DataResponse<Article>>()
            )
            .subscribe(object : Consumer<DataResponse<Article>> {
                @Throws(Exception::class)
                override fun accept(dataResponse: DataResponse<Article>) {
                    mView.loadMoreDataSuccess(dataResponse.data.datas)
                }
            }, object : Consumer<Throwable> {
                @Throws(Exception::class)
                override fun accept(throwable: Throwable) {
                    mView.showFailMsg(mContext.getString(R.string.paging_load) + throwable.message)
                }
            })
    }
}